addons/website_sale/controllers/main.py

```diff
@@ -27,11 +26,41 @@
 from odoo.addons.website.models.ir_http import sitemap_qs2dom
 from odoo.addons.website_sale.const import SHOP_PATH
 from odoo.addons.website_sale.models.website import (
-    PRICELIST_SESSION_CACHE_KEY,
     PRICELIST_SELECTED_SESSION_CACHE_KEY,
+    PRICELIST_SESSION_CACHE_KEY,
 )
 
 
+def handle_product_params_error(exc, product, category=None, **kwargs):
+    """ Handle access and missing errors related to product or category on the eCommerce.
+
+    This function is intended to prevent access-related exceptions when a user attempts to view a
+    product or category page. It checks if the provided product and category records still exist and
+    are accessible, and then attempts to redirect to a valid fallback route if possible. If no valid
+    route is found, it returns a 404 response code.
+
+    :param odoo.exceptions.AccessError | odoo.exceptions.MissingError exc: The exception thrown
+            by _check_access `base.models.ir_http._pre_dispatch`.
+    :param product.template product: The product the user is trying to access.
+    :param product.public.category category: The category the user is trying to access, if any.
+    :param dict kwargs: Optional data. This parameter is not used here.
+    :return: A redirect response to a valid shop or product page, or a 404 error code if no valid
+             fallback is found.
+    :rtype: int | Response
+    """
+    product = product.exists()
+    if category:
+        category = category.exists()
+
+    if category and not (product and product.sudo().is_published):
```

Comment:
```suggestion
    if category and not (product and product._has_access()):
```
`is_published` is our logic, but it might be modified in the `ir.rules`(you didn't consider `sale_ok` here either)